capsules/src/net/sixlowpan.rs

```diff
@@ -166,39 +285,42 @@ fn is_fragment(packet: &[u8]) -> bool {
     (mask == lowpan_frag::FRAGN_HDR) || (mask == lowpan_frag::FRAG1_HDR)
 }
 
-/// struct TxState
-/// --------------
-/// This struct tracks the per-client transmit state for a single IPv6 packet.
-/// The FragState struct maintains a list of TxState structs, sending each in
-/// order.
-pub struct TxState<'a> {
+/// Tracks the global transmit state for a single IPv6 packet.
+///
+/// Since transmit is serialized, the `Sixlowpan` struct only contains
+/// a reference to a single TxState (that is, we can only have a single
+/// outstanding transmission at the same time).
+///
+/// This struct maintains a reference to the full IPv6 packet, the source/dest
+/// MAC addresses and PanIDs, security/compression/fragmentation options,
+/// per-fragmentation state, and some global state.
+struct TxState {
+    // State for the current transmission
     packet: TakeCell<'static, [u8]>,
     src_pan: Cell<PanID>,
     dst_pan: Cell<PanID>,
     src_mac_addr: Cell<MacAddress>,
     dst_mac_addr: Cell<MacAddress>,
     security: Cell<Option<(SecurityLevel, KeyId)>>,
-    dgram_tag: Cell<u16>,
+    dgram_tag: Cell<u16>, // Used to identify particular fragment streams
     dgram_size: Cell<u16>,
     dgram_offset: Cell<usize>,
     fragment: Cell<bool>,
     compress: Cell<bool>,
-    client: Cell<Option<&'static TransmitClient>>,
 
-    next: ListLink<'a, TxState<'a>>,
-}
-
-impl<'a> ListNode<'a, TxState<'a>> for TxState<'a> {
-    fn next(&'a self) -> &'a ListLink<TxState<'a>> {
-        &self.next
-    }
+    // Global transmit state
+    tx_dgram_tag: Cell<u16>,
+    tx_busy: Cell<bool>,
+    tx_buf: TakeCell<'static, [u8]>,
 }
 
-impl<'a> TxState<'a> {
-    /// TxState::new
-    /// ------------
-    /// This constructs a new, default TxState struct.
-    pub fn new() -> TxState<'a> {
+impl TxState {
+    /// Creates a new `TxState`
+    ///
+    /// # Arguments
+    ///
+    /// `tx_buf` - A buffer for storing fragments the size of a 802.15.4 frame.
```

Comment:
How big does this buffer need to be? How large is a fragment size? Does it make sense for this to be encoded in the buffer type? e.g. `&'static mut [u8; 256]`

Reply:
I think that the fragment size is constant (~127 bytes, though I have to check), so it should be encoded in the buffer type.

Reply:
Especially since this is not going to be passed back and forth except with the 802.15.4 driver. If it ends up being a huge pain to do, no big deal, let's just document really clearly how large it needs to be

Reply:
I'm not quite sure there's a point to this. Eventually you'll still need to take a static mutable slice of it (because the radio returns buffers through the callback as static mutable slices, so we cannot store them as `&'static mut [u8; BLOCK_SIZE]` internally). The most you could do is make the constructor signature require a fixed-size buffer.

Reply:
Well then you would have to modify the crypto layer to only accept buffers of size 127.

Reply:
Oh ok, I understand. So either modifying just the constructor or at least documenting seems best.